File: libs/ui_pages_preview/src/qt/web_document_preview_widget.cpp
#include <pwb/ui_pages_preview/qt/web_document_preview_widget.hpp>

#include <QFileInfo>
#include <QLabel>
#include <QUrl>
#include <QVBoxLayout>

#if defined(PWB_UI_PAGES_PREVIEW_HAVE_WEBENGINE)
#include <QWebEnginePage>
#include <QWebEngineProfile>
#include <QWebEngineSettings>
#include <QWebEngineUrlRequestInfo>
#include <QWebEngineUrlRequestInterceptor>
#include <QWebEngineView>
#endif

#include <pwb/ui_pages_preview/preview_settings.hpp>
#include <pwb/ui_pages_preview/url_filter.hpp>

namespace pwb::ui_pages_preview {

#if defined(PWB_UI_PAGES_PREVIEW_HAVE_WEBENGINE)

namespace {

// _LocalOnlyRequestInterceptor: block WebEngine resource requests outside
// the local document sandbox {file, data, about, blob}.
class LocalOnlyRequestInterceptor : public QWebEngineUrlRequestInterceptor {
public:
    using QWebEngineUrlRequestInterceptor::QWebEngineUrlRequestInterceptor;

    void interceptRequest(QWebEngineUrlRequestInfo& info) override {
        if (!local_scheme_allowed(info.requestUrl().scheme().toStdString())) {
            info.block(true);
        }
    }
};

// _LocalOnlyPage: reject user-initiated navigation away from local
// document content.
class LocalOnlyPage : public QWebEnginePage {
public:
    using QWebEnginePage::QWebEnginePage;

    bool acceptNavigationRequest(const QUrl& url, NavigationType,
                                 bool isMainFrame) override {
        Q_UNUSED(isMainFrame);
        return local_scheme_allowed(url.scheme().toStdString());
    }
};

}  // namespace

#endif  // PWB_UI_PAGES_PREVIEW_HAVE_WEBENGINE

WebDocumentPreviewWidget::WebDocumentPreviewWidget(QWidget* parent)
    : QWidget(parent) {
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
#if defined(PWB_UI_PAGES_PREVIEW_HAVE_WEBENGINE)
    engine_view_ = new QWebEngineView(this);
    auto* profile = new QWebEngineProfile(engine_view_);
    auto* interceptor = new LocalOnlyRequestInterceptor(profile);
    profile->setUrlRequestInterceptor(interceptor);
    auto* page = new LocalOnlyPage(profile, engine_view_);
    engine_view_->setPage(page);
    engine_view_->settings()->setAttribute(
        QWebEngineSettings::LocalContentCanAccessRemoteUrls, false);
    layout->addWidget(engine_view_);
#else
    placeholder_ = new QLabel(QStringLiteral("Web 文档预览不可用"));
    placeholder_->setAlignment(Qt::AlignCenter);
    layout->addWidget(placeholder_);
#endif
}

void WebDocumentPreviewWidget::load_document(const QString& path,
                                             const QString& html) {
#if defined(PWB_UI_PAGES_PREVIEW_HAVE_WEBENGINE)
    const QUrl base_url =
        QUrl::fromLocalFile(QFileInfo(path).absolutePath() + QLatin1Char('/'));
    if (!html.isEmpty()) {
        engine_view_->setHtml(html, base_url);
    } else {
        engine_view_->load(QUrl::fromLocalFile(path));
    }
#else
    Q_UNUSED(path);
    Q_UNUSED(html);
#endif
}

void WebDocumentPreviewWidget::apply_settings(const PreviewSettings& settings) {
#if defined(PWB_UI_PAGES_PREVIEW_HAVE_WEBENGINE)
    engine_view_->setZoomFactor(settings.font_size / 12.0);
#else
    Q_UNUSED(settings);
#endif
}

}  // namespace pwb::ui_pages_preview

File: legacy/python_reference/product/paleo_workbench/ui/pages/web_document_preview_widget.py
from __future__ import annotations

from pathlib import Path

from PySide6.QtCore import QUrl
from PySide6.QtWidgets import QVBoxLayout, QWidget


class _LocalOnlyRequestInterceptor:
    """Block WebEngine resource requests outside the local document sandbox.

    Lazily inherits from QWebEngineUrlRequestInterceptor at import time to
    avoid forcing WebEngine initialization when preview_widgets is imported.
    """

    _ALLOWED_SCHEMES = {"file", "data", "about", "blob"}

    def interceptRequest(self, info) -> None:
        if info.requestUrl().scheme() not in self._ALLOWED_SCHEMES:
            info.block(True)


class _LocalOnlyPage:
    """Reject user-initiated navigation away from local document content.

    Lazily inherits from QWebEnginePage at construction time.
    """

    _ALLOWED_SCHEMES = {"file", "data", "about", "blob"}

    def acceptNavigationRequest(self, url, navigation_type, is_main_frame):
        del navigation_type, is_main_frame
        return url.scheme() in self._ALLOWED_SCHEMES


class WebDocumentPreviewWidget(QWidget):
    """Render local HTML or bounded Markdown output without network access.

    Inherits from QWidget (no WebEngine dependency at import time). The
    QWebEngineView is created lazily in __init__, so importing
    preview_widgets does not trigger WebEngine subprocess initialization.
    """

    def __init__(self, parent=None):
        super().__init__(parent)
        from PySide6.QtWebEngineCore import (
            QWebEnginePage,
            QWebEngineProfile,
            QWebEngineSettings,
            QWebEngineUrlRequestInterceptor,
        )
        from PySide6.QtWebEngineWidgets import QWebEngineView

        class _Interceptor(QWebEngineUrlRequestInterceptor, _LocalOnlyRequestInterceptor):
            pass

        class _Page(QWebEnginePage, _LocalOnlyPage):
            pass

        self._engine_view = QWebEngineView(self)
        self._profile = QWebEngineProfile(self)
        self._interceptor = _Interceptor(self._profile)
        self._profile.setUrlRequestInterceptor(self._interceptor)
        self._page = _Page(self._profile, self._engine_view)
        self._engine_view.setPage(self._page)
        self._engine_view.settings().setAttribute(
            QWebEngineSettings.WebAttribute.LocalContentCanAccessRemoteUrls,
            False,
        )
        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.addWidget(self._engine_view)

    def load_document(self, path: str, html: str = "") -> None:
        base_url = QUrl.fromLocalFile(str(Path(path).parent) + "/")
        if html:
            self._engine_view.setHtml(html, base_url)
        else:
            self._engine_view.load(QUrl.fromLocalFile(path))

    def apply_settings(self, settings) -> None:
        self._engine_view.setZoomFactor(settings.font_size / 12.0)
